### PR TITLE
add slot bitmap

### PR DESCRIPTION
**Pull request by johannbarbie**
resolves: https://github.com/leapdao/leap-contracts/issues/266

### CODE REVIEW

File: test/poaOperator.js [hunk 1]
```diff
@@ -78,48 +77,64 @@ contract('PoaOperator', (accounts) => {
...
+      it('should allow to set slot but prevent submission with CAS bitmap of 0/1', async () => {
         const data = await operator.contract.methods.setSlot(0, alice, alice).encodeABI();
         await proxy.applyProposal(data, {from: admin});
         const consensusRoot = '0x01';
         const casBitmap = '0x00'; // no bits set: 0000 0000
         await operator.submitPeriodWithCas(0, p[0], consensusRoot, casBitmap, { from: alice }).should.be.rejectedWith(EVMRevert);
       });
 
-      it('should prevent submission with CAS bitmap of 1/3', async () => {
+      it('should prevent submission with CAS bitmap of 1/1', async () => {
```
Comment (troggy) [inline test/poaOperator.js]: why? 1/1 is a valid situation IMO

Comment (johannbarbie) [reply test/poaOperator.js]: true

File: contracts/PoaOperator.sol [hunk 2]
```diff
@@ -99,6 +99,14 @@ contract PoaOperator is Adminable {
     _setSlot(_slotId, _signerAddr, _tenderAddr);
   }
 
+  function setTakenSlotBitmap(uint256 _before, uint256 _pos, bool isActive) internal pure returns (uint256) {
+    if (isActive) {
+      return _before | (0x01 << (255 - _pos));
+    } else {
+      return _before & (uint256(-2) << (255 - _pos));
```
Comment (troggy) [inline contracts/PoaOperator.sol]:
```suggestion
      // uint256(-2) is 0xFFFFF....FFE (the very last bit off)
      return _before & (uint256(-2) << (255 - _pos));
```

File: contracts/PoaOperator.sol [hunk 3]
```diff
@@ -123,6 +132,7 @@ contract PoaOperator is Adminable {
     if (_signerAddr == address(0) && _tenderAddr == 0) {
       slot.activationEpoch = uint32(lastCompleteEpoch + 3);
       slot.eventCounter++;
+      takenSlots = setTakenSlotBitmap(takenSlots, _slotId, false);
```
Comment (troggy) [inline contracts/PoaOperator.sol]: Can we have some unit tests checking taken slots bitmap is set correctly after logout?

E.g. this test fails for me (I've temporary made `setTakenSlotBitmap` public):

```js
const bits110 = '0x0000000000000000000000000000000000000000000000000000000000000006';
assert(
  await operator.setTakenSlotBitmap(bits110, 253, false)
).equal(BN(2));
```

while 

```js
assert(
  await operator.setTakenSlotBitmap(bits110, 254, false)
).equal(BN(4));
```

works

Comment (johannbarbie) [reply contracts/PoaOperator.sol]: checking ...

Comment (johannbarbie) [reply contracts/PoaOperator.sol]: @troggy great catch 🎉 logic was broken, deleting all bits on the right, as shifting filled up the bitmask with 0s.

File: contracts/PoaOperator.sol [hunk 4]
```diff
@@ -99,6 +99,17 @@ contract PoaOperator is Adminable {
     _setSlot(_slotId, _signerAddr, _tenderAddr);
   }
 
+  function setTakenSlotBitmap(uint256 _before, uint256 _pos, bool isActive) internal pure returns (uint256) {
+    if (isActive) {
+      return _before | (0x01 << (255 - _pos));
+    } else {
+      // if the bit is on, turn it off through XOR
+      if (_before | (0x01 << (255 - _pos)) > 0) {
```
Comment (troggy) [inline contracts/PoaOperator.sol]: can be replaced with

```
return _before & ~(0x01 << (255 - _pos));
```

?

Comment (troggy) [reply contracts/PoaOperator.sol]: 1. (0x01 << (255 - _pos) — sets the bit e.g. (0x000100)
2. ~... — inverts the number so the bit is off (e.g. 0x111011)
3. .. &  .. — clears the bit in `_before`

**Pull request merged**